src/Microsoft.ML.Featurizers/DateTimeTransformer.cs

```diff
@@ -0,0 +1,782 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Security;
+using System.Text;
+using Microsoft.ML;
+using Microsoft.ML.CommandLine;
+using Microsoft.ML.Data;
+using Microsoft.ML.EntryPoints;
+using Microsoft.ML.Featurizers;
+using Microsoft.ML.Internal.Utilities;
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Transforms;
+using Microsoft.Win32.SafeHandles;
+using static Microsoft.ML.Featurizers.CommonExtensions;
+
+[assembly: LoadableClass(typeof(DateTimeTransformer), null, typeof(SignatureLoadModel),
+    DateTimeTransformer.UserName, DateTimeTransformer.LoaderSignature)]
+
+[assembly: LoadableClass(typeof(IRowMapper), typeof(DateTimeTransformer), null, typeof(SignatureLoadRowMapper),
+   DateTimeTransformer.UserName, DateTimeTransformer.LoaderSignature)]
+
+[assembly: EntryPointModule(typeof(DateTimeTransformerEntrypoint))]
+
+namespace Microsoft.ML.Featurizers
+{
+
+    public static class DateTimeTransformerExtensionClass
+    {
+        /// <summary>
+        /// Create a <see cref="DateTimeTransformerEstimator"/>, which splits up the input column specified by <paramref name="inputColumnName"/>
+        /// into all its individual datetime components. Input column must be of type Int64 representing the number of seconds since the unix epoc.
+        /// This transformer will append the <paramref name="columnPrefix"/> to all the output columns. If <paramref name="columnsToDrop"/> is empty,
+        /// then all the columns are returned. Otherwise, the columns listed in the array will be dropped from the return value.
+        /// </summary>
+        /// <param name="catalog">Transform catalog</param>
+        /// <param name="inputColumnName">Input column name</param>
+        /// <param name="columnPrefix">Prefix to add to the generated columns</param>
+        /// <param name="columnsToDrop">List of columns to drop, if any</param>
```

Comment:
One drawback to using `columnsToDrop` is that if a user is only interested in the `DayOfYear` and `DayOfWeekLabel`, they need to explicitly exclude all the rest. And then if we ever add a new output column in a future release, they would start getting it, and would need to add the new column to the "to drop" list.

Reply:
I might recommend a bitwise use of an enum (flags) to say which columns to include.

Default would be `DateTimeTransformerEstimator.ColumnsProduced.All = 0xFFFFFFFFFFFFFFFF` (uint64).

Then users can bitwise `or`/`and-not` to select the columns they want to keep/drop.

## Example usage
All output except `DayOfYear`: (subtractive column selection)
```c#
var pipeline = mlContext.Transforms.DateTimeTransformer("Date", "DTC", 
    DateTimeTransformerEstimator.ColumnsProduced.All & ~DateTimeTransformerEstimator.ColumnsProduced.DayOfYear)
```

Only `DayOfYear` and `DayOfWeekLabel`: (additive column selection)
```c#
var pipeline = mlContext.Transforms.DateTimeTransformer("Date", "DTC", 
    DateTimeTransformerEstimator.ColumnsProduced.DayOfYear | DateTimeTransformerEstimator.ColumnsProduced.DayOfWeekLabel)
```

Reply:
Apparently, there's a `[Flags]` attribute in C# -- https://docs.microsoft.com/en-us/dotnet/api/system.flagsattribute?view=netcore-3.0

```c#
[Flags] public enum ColumnsProduced : uint64
{
    Year = 1 << 0,
    Month = 1 << 1,
    Day = 1 << 2,
    Hour = 1 << 3, 
    ...
}
             WeekOfMonth, QuarterOfYear,
```

Reply:
Do you have a way you think would be better? We had also thought about not even including that ability here and just letting the user use the `DropColumns` transformer to remove any columns they didn't want. That would keep this code cleaner and not duplicate that type of functionality. Thoughts on that?

---
In reply to: [353911746](https://github.com/dotnet/machinelearning/pull/4521#discussion_r353911746) [](ancestors = 353911746)

Reply:
I dont think that doing flags is a great idea personally. I think enums are more clear (if more verbose) than flags usually. As I mentioned to Eric in this same thread though, we can remove this capability from here and just let the user use the `DropColumn` transformer to remove anything they dont want. What is your opinion on that?

---
In reply to: [353968394](https://github.com/dotnet/machinelearning/pull/4521#discussion_r353968394) [](ancestors = 353968394)

Reply:
> We had also thought about not even including that ability here and just letting the user use the DropColumns transformer to remove any columns they didn't want.

The way [IDataView works](https://github.com/dotnet/machinelearning/blob/master/docs/code/IDataViewDesignPrinciples.md#lazy-computation-and-active-columns) in ML.NET is that if a column isn't used downstream, it isn't even materialized. So if a data transform produces 100 columns, but only 10 columns are used by the learner, the other 90 columns don't get materialized. See the linked doc for more info.

So if you are open to not even including the `drop` ability here, I think that works and the perf should be the same.

Reply:
I am open to that. I already synced with my team on it and they are fine with that as well. I will go ahead then and remove it.

---
In reply to: [354479281](https://github.com/dotnet/machinelearning/pull/4521#discussion_r354479281) [](ancestors = 354479281)

Reply:
> The way IDataView works in ML.NET is that if a column isn't used downstream, it isn't even materialized. So if a data transform produces 100 columns, but only 10 columns are used by the learner, the other 90 columns don't get materialized. See the linked doc for more info.

While true in general, I expect this featurizer isn't aware of which columns are being requested of it, and will produce all columns.

Reply:
I think the better approach is to use the DropColumns facility that is already present in ML.NET. 

---
In reply to: [354480672](https://github.com/dotnet/machinelearning/pull/4521#discussion_r354480672) [](ancestors = 354480672,354479281)

Reply:
>  I expect this featurizer isn't aware of which columns are being requested of it

Maybe it should be made aware?

Reply:
I like the usability of the flag method. 

There is a user need of not including all outputs as overfitting occurs mainly on the larger time scales, like year number. Including the year in a classification or regression model often leads to the model memorizing the label value specifically for a date range. 

For instance the model will memorize that 2017-03-01 to 2017-04-05, the label value was high. This information is not valuable to memorize when predicting future data. Instead we want to ensure it memorizes that traffic volumes (the label) are higher on weekends.

Hence it nice to encourage users (as part of the timedate featurizer API itself) to think about which columns to include/exclude. This is as opposed to a user needing a separate DropColumns step in their pipeline.

Reply:
So so far we have had 3 suggestions.
1 - Dont support dropping columns as a native part of this transformer. 
2 - Use Enums.
3 - Use Flags.

Can we finalize on the decision? I personally agree with Harish that since we already have a drop columns transformer that we can just use that and not have that functionality be present here.

Reply:
I'd recommend a way to proactively select the output columns to create during the setup of the transform. Instead of dropping later, just don't create them.

Generally, we should assume users won't think (or know) to drop later. For usability, placing it in the constructor exposes users to the idea, and forces them to consider which columns will be beneficial.

Reply:
For now, since this functionality already exists inside of ML.NET as a separate transformer we will be leaving this as is. We can revisit this later as the need arises. Even if this transformer produces all the columns, if no downstream transformers request them then they will never be realized. It shouldn't cause any performance impact this way.